velox/common/caching/SsdFile.h

```diff
@@ -0,0 +1,297 @@
+// backed by each SsdFile is selected on a hash of the storage file
+// number of the cached data. Each file consists of an integer number
+// of 64MB regions. Each region has a pin count and an read
+// count. Cache replacement takes place region by region, preferring
+// regions with a smaller read count. Entries do not span
+// regions. Otherwise entries are consecutive byte ranges inside
+// their region.
+class SsdFile {
+ public:
+  static constexpr uint64_t kRegionSize = 1 << 26; // 64MB
+
+  // Constructs a cache backed by filename. Discards any previous
+  // contents of filename.
+  SsdFile(const std::string& filename, int32_t shardId, int32_t maxRegions);
+
+  // Adds entries of  'pins'  to this file. 'pins' must be in read mode and
+  // those pins that are successfully added to SSD are marked as being on SSD.
+  // The file of the entries must be a file that is backed by 'this'.
+  void write(std::vector<CachePin>& pins);
+
+  // Finds an entry for 'key'. If no entry is found, the returned pin is empty.
+  SsdPin find(RawFileCacheKey key);
+
+  // Copies the data at 'run' into 'entry'. Checks that the entry
+  // and run sizes match. The quantization of SSD cache matches that
+  // of the memory cache.
+  void load(SsdRun run, AsyncDataCacheEntry& entry);
+
+  // Reads the backing file as per ReadFile::preadv(). 'numEntries' is
+  // used only for updating statistics. Allows implementing coalesced
+  // read of nearby entries.
+  void read(
+      uint64_t offset,
+      const std::vector<folly::Range<char*>> buffers,
+      int32_t numEntries);
+
+  // Increments the pin count of the region of 'offset'.
+  void pinRegion(uint64_t offset);
+
+  // Decrements the pin count of the region of 'offset'. If the pin
+  // count goes to zero and evict is due, starts the evict.
+  void unpinRegion(uint64_t offset);
+
+  // Asserts that the region of 'offset' is pinned. This is called by
+  // the pin holder. The pin count can be read without mutex.
+  void checkPinned(uint64_t offset) const {
+    VELOX_CHECK_LT(0, regionPins_[regionIndex(offset)]);
+  }
+
+  // Returns the region number corresponding to offset.
+  static int32_t regionIndex(uint64_t offset) {
+    return offset / kRegionSize;
+  }
+
+  // Updates the read count of a region.
+  void regionUsed(int32_t region, int32_t size) {
+    regionScore_[region] += size;
+  }
+
+  int32_t maxRegions() const {
+    return maxRegions_;
+  }
+
+  int32_t shardId() const {
+    return shardId_;
+  }
+
+  // Adds 'stats_' to 'stats'.
+  void updateStats(SsdCacheStats& stats) const;
+
+  // Resets this' to a post-construction empty state. See SsdCache::clear().
+  void clear();
+
+ private:
+  static constexpr int32_t kDecayInterval = 1000;
+
+  // Increments the pin count of the region of 'offset'. Caller must hold
+  // 'mutex_'.
+  void pinRegionLocked(uint64_t offset) {
+    ++regionPins_[regionIndex(offset)];
+  }
+
+  // Returns [start, size] of contiguous space for storing data of a
```

Comment:
> Returns [start, size]

Does 'start' here means offset in the file? Would it make sense to use SsdRun instead of std::pair<uint64_t, int32_t> ? This method could return std::optional<SsdRun> and allow for indicating no-space using std::nullopt.

Reply:
Looks like this comment got lost. Repeating. "Does 'start' here means offset in the file? Would it make sense to use SsdRun instead of std::pair<uint64_t, int32_t> ?"